services/command-router/src/main/java/org/eclipse/hono/commandrouter/impl/CommandProcessingQueue.java

```diff
@@ -0,0 +1,69 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.commandrouter.impl;
+
+import java.util.function.Supplier;
+
+import org.eclipse.hono.client.command.CommandContext;
+
+import io.vertx.core.Future;
+
+/**
+ * Queue with the commands currently being processed.
+ * <p>
+ * The final step of processing a command, forwarding it to its target, is invoked here maintaining the
+ * order of the incoming commands.
+ *
+ * @param <T> The type of command context of the commands added here.
+ */
+public interface CommandProcessingQueue<T extends CommandContext> {
+
+    /**
+     * Adds the command represented by the given command context.
+     *
+     * @param commandContext The context containing the command to add.
+     * @throws NullPointerException if commandContext is {@code null}.
+     */
+    void add(T commandContext);
+
+    /**
+     * Removes the command represented by the given command context.
+     * <p>
+     * To be used for commands for which processing resulted in an error
+     * and {@link #applySendCommandAction(CommandContext, Supplier)}
+     * will not be invoked.
+     *
+     * @param commandContext The context containing the command to remove.
+     * @return {@code true} if the command was removed.
+     */
+    boolean remove(T commandContext);
+
+    /**
+     * Either invokes the given sendAction directly if it is next-in-line or makes sure the action is
+     * invoked at a later point in time according to the order in which commands were originally received.
```

Comment:
is there a particular reason why we need a separate method for this? It looks like the *add* method could do the very same thing, if it would also accept the send action supplier. The question seems to be: who is responsible for sending the command? The queue (implementation) or the client code using the queue?

Reply:
Do you mean changing:
```
// command just arrived
commandQueue.add(commandContext);

// ... steps to determine the command target ...
// command now ready to be forwarded (but another might need to be forwarded first)
commandQueue.applySendCommandAction(commandContext, sendActionSupplier);
```
to something like:
```
// command just arrived
commandQueue.add(commandContext, sendActionSupplier);

// ... steps to determine the command target ...
// command now ready to be forwarded (but another command might need to be forwarded first)
sendActionSupplier.setTarget(target);
commandQueue.markAsReadyToBeSent(commandContext); // triggering associated sendActionSupplier when appropriate
```
?

In general, there are still 2 interactions with the queue needed here. First, adding the command to the queue, defining the order in which to later forward the command. And second, marking the point in time when a command is ready to be sent (potentially having to wait for another command to be forwarded first).

> The question seems to be: who is responsible for sending the command? The queue (implementation) or the client code using the queue?

The queue implementation knows _when_ a particular command can be sent. So, the queue implementation needs to trigger sending the command.

Reply:
Can't the queue implementation also do the
> ... steps to determine the command target ...

part?

Reply:
You mean in the sense of letting the queue invoke a `getCommandTarget` supplier, provided as a parameter of the queue `add` method?
I don't necessarily see a benefit of such a refactoring.

I mean, I could imagine having only a single 
`queue.add(commandContext, firstProcessingStepSupplier, orderedLastProcessingStepSupplier)` 
method (1st processing step=getting target, last step=sending message).
This would involve letting `orderedLastProcessingStepSupplier` work with the `firstProcessingStepSupplier` result (i.e. the found command target).
I don't know if that would make the overall code more readable. In any case, I would rather make such a refactoring in a separate PR.

Reply:
i had been hoping that we would be able to encapsulate the processing of the queued commands within the queue. That way we would have higher cohesion and encapsulation. But if that is not feasible/possible then we should stick with what you got ..

Reply:
For this PR, I would suggest sticking with the existing approach. 
I'll have a look into whether/how to refactor it further then.

Reply:
Ok